File: DAA/Assignment 5/ass_5.py
import time

def print_board(board, n):
    for i in range(n):
        row = ""
        for j in range(n):
            row += "Q " if board[i][j] else ". "
        print(row)
    print()

def is_safe(board, row, col, n):
    # Check this column on upper rows
    for i in range(row):
        if board[i][col]:
            return False
    
    # Check upper diagonal on left side
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j]:
            return False

    # Check upper diagonal on right side
    for i, j in zip(range(row, -1, -1), range(col, n)):
        if board[i][j]:
            return False

    return True

def solve_nqueens(board, row, n):
    if row >= n:
        print_board(board, n)
        return True

    for col in range(n):
        if is_safe(board, row, col, n):
            board[row][col] = 1  # Place queen
            if solve_nqueens(board, row + 1, n):
                return True
            board[row][col] = 0  # Backtrack if placing the queen leads to no solution

    return False

def nqueens(n):
    board = [[0] * n for _ in range(n)]
    # Place the first queen at a given position (first row, first column for example)
    board[0][0] = 1
    if not solve_nqueens(board, 1, n):
        print("No solution found")
    return board

# Example usage
n = 8
start_time = time.time()
nqueens(n)
end_time = time.time()

print(f"Execution Time: {end_time - start_time:.6f} seconds")

File: DAA/Assignment 5/ass_5.cpp
#include <iostream>
#include <vector>
#include <chrono>
using namespace std;

void printBoard(const vector<vector<int>>& board, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cout << (board[i][j] ? "Q " : ". ");
        }
        cout << endl;
    }
    cout << endl;
}

bool isSafe(const vector<vector<int>>& board, int row, int col, int n) {
    // Check this column on upper rows
    for (int i = 0; i < row; i++) {
        if (board[i][col])
            return false;
    }

    // Check upper diagonal on left side
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j])
            return false;
    }

    // Check upper diagonal on right side
    for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
        if (board[i][j])
            return false;
    }

    return true;
}

bool solveNQueens(vector<vector<int>>& board, int row, int n) {
    if (row >= n) {
        printBoard(board, n);
        return true;
    }

    for (int col = 0; col < n; col++) {
        if (isSafe(board, row, col, n)) {
            board[row][col] = 1;  // Place queen
            if (solveNQueens(board, row + 1, n))
                return true;
            board[row][col] = 0;  // Backtrack
        }
    }

    return false;
}

void nQueens(int n) {
    vector<vector<int>> board(n, vector<int>(n, 0));
    // Place the first queen at a given position (first row, first column)
    board[0][0] = 1;
    if (!solveNQueens(board, 1, n)) {
        cout << "No solution found" << endl;
    }
}

int main() {
    int n = 8;
    
    auto start = chrono::high_resolution_clock::now();
    nQueens(n);
    auto end = chrono::high_resolution_clock::now();
    
    chrono::duration<double> execution_time = end - start;
    cout << "Execution Time: " << execution_time.count() << " seconds" << endl;

    return 0;
}
